Added implementation on the python3

lubchenko05 opened the pull request.

lubchenko05 committed:
Added implementation on the python3

agil commented:
1. Чрезмерное использование двойного (**__max_listeners**, **__events**, **__events_once**) и одинарного (**_flag_l**, **_flag_e**) подчеркивания (см. https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables)

2. методы **add_listener** и **on** абсолютно одинаковые и вообще много повторяющегося кода, его нужно выносить в методы и использовать

3. Для **__events** и **__events_once** было бы удобнее использовать defaultdict, если это разрешено, либо dict.setdefault

4. Многие методы можно переписать короче и более читаемо используя dict.get
```python
def listeners(self, event: str) -> List[Callable]:
    return self.__events.get(event, []) + self.__events_once.get(event, [])

def listener_count(self, event: str) -> int:
    return len(self.listeners(event))

def emit(self, event: str, *args, **kwargs) -> bool:
  listeners = self.__events.get(event, []) + self.__events_once.pop(event, [])
  if not listeners:
    return False
  
  for listener in listeners:
    listener(*args, **kwargs)
```

5. **__max_listeners** не может быть меньше 0, но почему-то повсюду сравнивается с -1

6. соответственно методы можно переписать проще:
```python
def on(self, event: str, listener: Callable) -> 'EventEmitter':
    if self.listener_count(event) >= self.__max_listeners:
        raise Exception('Exceeded ...')
    self.__events.setdefault(event, []).append(listener)

def prepend_listener(self, event: str, listener: Callable) -> 'EventEmitter':
    if self.listener_count(event) >= self.__max_listeners:
        raise Exception('Exceeded ...')
    self.__events.setdefault(event, []).insert(0, listener)

def prepend_once_listener(self, event: str, listener: Callable) -> 'EventEmitter':
    if self.listener_count(event) >= self.__max_listeners:
        raise Exception('Exceeded ...')
    self.__events_once.setdefault(event, []).insert(0, listener)
```

7.  Zen of Python: **"flat is better than nested"**
```python 
def remove_listener(self, event: str, listener: Callable) -> 'EventEmitter':
  if listener not in self.listeners():
    raise Exception('Listener was not found')

  if listener in self.__events.get(event, []):
    self.__events[event].remove(listener)

  if listener in self.events_once.get(event, []):
    self.__events_once[event].remove(listener)
    

def remove_all_listener(self, event: str=None) -> 'EventEmitter':
  if not self.listeners(event):
    raise Exception('Event was not found with name - %s' % event)

  if event in self.__events:
    del self.__events[event]

  if event in self.__events_once:
    del self.__events_once[event]

```

8. Ну и dict.keys лучше чем list comprehension с лишними переменными + вероятно нужно убирать дубликаты
```python
def event_names(self) -> List[str]:
    return list(set(self.__events.keys()) + set(self.__events_once.keys()))
```

lubchenko05 commented:
1. _ и __ я использую для того что бы объявлять приватные и протектед
переменные.
2. Это сделано для удобства и совместимости. Эти методы такие же как и в
оригинальном Event Emitter написанном на Node.js.
Примером может служить subscribe и forEach в Observable.
Согласен что копирование кода не есть правильным. Возможно стоит вызывать 1й через 2й.
5. Ставлю __max_listeners в -1 что бы сделать неограниченное количество listers.
С остальным согласен, переделаю.

agil commented:
1. Строго говоря в питоне нет никаких private/protected свойств. Особенно если это учебный пример, то имхо, не стоит вводить людей в заблуждение. _**Follow the pythonic way, Luke.**_ http://python.net/~goodger/projects/pycon/2007/idiomatic/handout.html#naming
1.1  Одинарное подчеркивание для локальных переменных внутри функции тоже избыточно
2. можно обойтись присвоением
```python 
def on(self):
    # code here
add_listener = on
```
3. конечно, но если посмотреть на реализацию set_max_listeners, то становится ясно, что такое присвоение не слишком возможно. А если дополнительно посмотреть в [документацию nod'овского eventEmitter'a ](https://nodejs.org/api/events.html#events_emitter_setmaxlisteners_n), то еще и поведение должно быть другим `The value can be set to Infinity (or 0) to indicate an unlimited number of listeners.`

lubchenko05 commented:
Да, еще в 8 пункте вы зачем-то делаете set из ключей одного словаря+сет из ключей второго, а ключи словаря и так уникальны там скорее должно быть:
```python
def event_names(self) -> List[str]:
    return list({*self.__events.keys(), *self.__events_once.keys()})
```

agil commented:
Да, ошибся, конечно. Тот пример вообще не рабочий, там должна быть операция **|** вместо **+**. А задача, конечно, сделать общий set из ключей events и events_once
```python
list(self._events.keys() | self._events_once.keys())
```

lubchenko05 committed:
Merge branch 'master' of https://github.com/HowProgrammingWorks/EventEmitter

lubchenko05 committed:
refactor code

lubchenko05 committed:
remove .idea

lubchenko05 committed:
refactor event_name

lubchenko05 committed:
add .gitignore

agil committed:
a bit more consistent with node implementation

use ValueError because it's more specific 
remove_listener and remove_all_listeners should not raise exceptions if event/listener does not exist
checking types is not pythonic way and should be avoided when possible
emit should return True if atleast one listener was called

agil reviewed (commented):
LGTM

agil commented:
могут, но это ближе к нодовскому поведению, которое говорит удалять по одному за раз. https://nodejs.org/api/events.html#events_emitter_removelistener_eventname_listener

> removeListener will remove, at most, one instance of a listener from the listener array. If any single listener has been added multiple times to the listener array for the specified eventName, then removeListener must be called multiple times to remove each instance.

строго говоря, чтоб совсем соответствовать, надо бы удалять (да и вызывать) обработчики строго в порядке добавления, но это нужно все менять и делать один общий список обработчиков

aqrln commented on Python/ee.py:
```diff
@@ -0,0 +1,123 @@
+        """
+        Set max quantity of listeners of the event
+        """
+        if int(n) < 0:
```
Do we really need to convert `n` to an `int` here?

aqrln commented on Python/readme.md:
```diff
@@ -0,0 +1,18 @@
+
+ee = EventEmitter()
+
+ee.on('operations', lambda x, y: print(str(x+y)))
```
You don't need to call `str()` explicitly here.

lubchenko05 committed:
Remove unnecessary type conversion. Fix bug with access to the field __max_listeners.

lubchenko05 commented:
@aqrln You're right. Fixed.

aqrln commented on Python/ee.py:
```diff
@@ -85,10 +91,13 @@ def set_max_listeners(self, n: int) -> None:
             raise ValueError('"n" argument must be a positive number')
         self.__max_listeners = int(n)
 
+    def get_max_listeners(self) -> int:
```
The name of this function is the same as of the one below.

lubchenko05 committed:
Remove duplicate method

aqrln commented on Python/.vscode/tags:
```diff
@@ -0,0 +1,26 @@
+!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
```
This looks like a stray file.

lubchenko05 commented:
Fixed

lubchenko05 commented:
@agil Hello, please review and approve changes
